Integration Tests for Quarkus Native Image for AWS Lambda do not work with multiple lambdas

### Describe the bug

I was running integration tests with quarkus in order to test that the native image actually works as intended, since I had issues before with forgetting to add some `@RegisterForReflection`. Our project contains multiple Lambda handlers which we test by having a `QuarkusTestProfile` overriding `quarkus.lambda.handler` for the respective lambda we want to test in a specific test.

When having multiple lambdas in the same project, as is generated by the mvn archetype described in the [documentation](https://quarkus.io/guides/amazon-lambda) it seems to be not trivially possible to run integration tests against multiple lambdas at the same time. Thus we resorted to the `QuarkusTestProfile`.

During the execution of the integration tests it seems that the test framework tries to start the native image multiple times. It ignores any port indication I might provide however:

```
Executing "target/test-1.0-SNAPSHOT-runner -Dquarkus.log.file.path=target/quarkus.log -Dquarkus.log.file.enable=true -Dquarkus-internal.aws-lambda.test-api=localhost:8081/_lambda_ -Dquarkus.lambda.mock-event-server.test-port=8082 -Dquarkus.configuration.build-time-mismatch-at-runtime=fail -Dquarkus.lambda.handler=first -Dquarkus.lambda.mock-event-server.dev-port=8082 -Dquarkus.http.port=8082 -Dquarkus.http.test-port=8082"
```

As you can see I tried overriding all ports, including `quarkus-internal.aws-lambda.test-api`. The `quarkus-internal.aws-lambda.test-api` is ignored however and set hard to port 8081.

Regular tests run fine.

### Expected behavior

Integration tests should run without issue, either be not ignoring the ports I set or by stopping the running image (or port listen) before starting the next.

### Actual behavior

Test crash because port is already in use:

```
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 3.595 s <<< FAILURE! - in org.test.SecondLambdaHandlerIT
[ERROR] org.test.SecondLambdaHandlerIT.testSimpleLambdaSuccess  Time elapsed: 0.001 s  <<< ERROR!
java.lang.RuntimeException:
java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkus.amazon.lambda.deployment.DevServicesLambdaProcessor#startEventServer threw an exception: java.lang.RuntimeException: java.util.concurrent.ExecutionException: java.net.BindException: Address already in use
...
```

### How to Reproduce?

[project.zip](https://github.com/quarkusio/quarkus/files/8521206/project.zip)

Open the project zip and run `mvn clean verify -Dnative -Dquarkus.native.container-build=true`



### Output of `uname -a` or `ver`

Linux NB01202 5.10.16.3-microsoft-standard-WSL2 #1 SMP Fri Apr 2 22:23:49 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "17.0.2" 2022-01-18

### GraalVM version (if different from Java)

-

### Quarkus version or git rev

2.8.0

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)

### Additional information

_No response_

--- a/extensions/amazon-lambda-http/runtime/src/main/java/io/quarkus/amazon/lambda/http/LambdaHttpHandler.java
+++ b/extensions/amazon-lambda-http/runtime/src/main/java/io/quarkus/amazon/lambda/http/LambdaHttpHandler.java
@@ -115,7 +115,12 @@ public class LambdaHttpHandler implements RequestHandler<APIGatewayV2HTTPEvent, 
                         }
                         final StringBuilder sb = new StringBuilder();
                         for (Iterator<String> valueIterator = allForName.iterator(); valueIterator.hasNext();) {
-                            sb.append(valueIterator.next());
+                            String val = valueIterator.next();
+                            if (name.equalsIgnoreCase("Transfer-Encoding")
+                                    && val.equals("chunked")) {
+                                continue; // ignore transfer encoding, chunked screws up message and response
+                            }
+                            sb.append(val);
                             if (valueIterator.hasNext()) {
                                 sb.append(",");
                             }

--- a/extensions/amazon-lambda-rest/runtime/src/main/java/io/quarkus/amazon/lambda/http/LambdaHttpHandler.java
+++ b/extensions/amazon-lambda-rest/runtime/src/main/java/io/quarkus/amazon/lambda/http/LambdaHttpHandler.java
@@ -97,7 +97,13 @@ public class LambdaHttpHandler implements RequestHandler<AwsProxyRequest, AwsPro
                     }
                     responseBuilder.setMultiValueHeaders(new Headers());
                     for (String name : res.headers().names()) {
+                        if (name.equalsIgnoreCase("Transfer-Encoding")) {
+                            continue; // ignore transfer encoding, chunked screws up message and response
+                        }
                         for (String v : res.headers().getAll(name)) {
+                            if (name.equalsIgnoreCase("Transfer-Encoding") && v.contains("chunked")) {
+                                continue;
+                            }
                             responseBuilder.getMultiValueHeaders().add(name, v);
                         }
                     }

--- a/extensions/amazon-lambda/common-deployment/src/main/java/io/quarkus/amazon/lambda/deployment/DevServicesLambdaProcessor.java
+++ b/extensions/amazon-lambda/common-deployment/src/main/java/io/quarkus/amazon/lambda/deployment/DevServicesLambdaProcessor.java
@@ -12,17 +12,17 @@ import org.jboss.logging.Logger;
 import io.quarkus.amazon.lambda.runtime.AmazonLambdaApi;
 import io.quarkus.amazon.lambda.runtime.LambdaHotReplacementRecorder;
 import io.quarkus.amazon.lambda.runtime.MockEventServer;
-import io.quarkus.bootstrap.classloading.QuarkusClassLoader;
 import io.quarkus.deployment.Feature;
 import io.quarkus.deployment.IsNormal;
 import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
 import io.quarkus.deployment.annotations.Produce;
 import io.quarkus.deployment.annotations.Record;
+import io.quarkus.deployment.builditem.CuratedApplicationShutdownBuildItem;
 import io.quarkus.deployment.builditem.DevServicesResultBuildItem;
 import io.quarkus.deployment.builditem.LaunchModeBuildItem;
-import io.quarkus.deployment.builditem.RuntimeApplicationShutdownBuildItem;
 import io.quarkus.deployment.builditem.ServiceStartBuildItem;
+import io.quarkus.deployment.dev.devservices.GlobalDevServicesConfig;
 import io.quarkus.runtime.LaunchMode;
 
 public class DevServicesLambdaProcessor {
@@ -53,12 +53,12 @@ public class DevServicesLambdaProcessor {
     }
 
     @Produce(ServiceStartBuildItem.class)
-    @BuildStep(onlyIfNot = IsNormal.class)
+    @BuildStep(onlyIfNot = IsNormal.class, onlyIf = GlobalDevServicesConfig.Enabled.class)
     public void startEventServer(LaunchModeBuildItem launchMode,
             LambdaConfig config,
             Optional<EventServerOverrideBuildItem> override,
             BuildProducer<DevServicesResultBuildItem> devServicePropertiesProducer,
-            BuildProducer<RuntimeApplicationShutdownBuildItem> runtimeApplicationShutdownBuildItemBuildProducer)
+            CuratedApplicationShutdownBuildItem closeBuildItem)
             throws Exception {
         if (!launchMode.getLaunchMode().isDevOrTest())
             return;
@@ -106,10 +106,6 @@ public class DevServicesLambdaProcessor {
             startMode = null;
             server = null;
         };
-        QuarkusClassLoader cl = (QuarkusClassLoader) Thread.currentThread().getContextClassLoader();
-        ((QuarkusClassLoader) cl.parent()).addCloseTask(closeTask);
-        if (launchMode.isTest()) {
-            runtimeApplicationShutdownBuildItemBuildProducer.produce(new RuntimeApplicationShutdownBuildItem(closeTask));
-        }
+        closeBuildItem.addCloseTask(closeTask, true);
     }
 }

--- a/extensions/azure-functions-http/runtime/src/main/java/io/quarkus/azure/functions/resteasy/runtime/BaseFunction.java
+++ b/extensions/azure-functions-http/runtime/src/main/java/io/quarkus/azure/functions/resteasy/runtime/BaseFunction.java
@@ -104,6 +104,11 @@ public class BaseFunction {
                     HttpResponse res = (HttpResponse) msg;
                     responseBuilder = request.createResponseBuilder(HttpStatus.valueOf(res.status().code()));
                     for (Map.Entry<String, String> entry : res.headers()) {
+                        if (entry.getKey().equalsIgnoreCase("Transfer-Encoding")
+                                && entry.getValue().contains("chunked")) {
+                            continue; // ignore transfer encoding, chunked screws up message and response
+                        }
+                        //log.info("header(" + entry.getKey() + ")=" + entry.getValue());
                         responseBuilder.header(entry.getKey(), entry.getValue());
                     }
                 }

